server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java

```diff
@@ -129,48 +126,56 @@ private void innerRun() throws Exception {
-                    () -> moveToNextPhase(fetchResults.getAtomicArray(), reducedQueryPhase),
-                    context
-                );
-                for (int i = 0; i < docIdsToLoad.length; i++) {
-                    List<Integer> entry = docIdsToLoad[i];
-                    RankDocShardInfo rankDocs = rankDocsPerShard == null || rankDocsPerShard.get(i).isEmpty()
-                        ? null
-                        : new RankDocShardInfo(rankDocsPerShard.get(i));
-                    SearchPhaseResult shardPhaseResult = searchPhaseShardResults.get(i);
-                    if (entry == null) { // no results for this shard ID
-                        if (shardPhaseResult != null) {
-                            // if we got some hits from this shard we have to release the context there
-                            // we do this as we go since it will free up resources and passing on the request on the
-                            // transport layer is cheap.
-                            releaseIrrelevantSearchContext(shardPhaseResult, context);
-                            progressListener.notifyFetchResult(i);
-                        }
-                        // in any case we count down this result since we don't talk to this shard anymore
-                        counter.countDown();
-                    } else {
-                        executeFetch(
-                            shardPhaseResult,
-                            counter,
-                            entry,
-                            rankDocs,
-                            (lastEmittedDocPerShard != null) ? lastEmittedDocPerShard[i] : null
-                        );
-                    }
+                innerRunFetch(scoreDocs, numShards, reducedQueryPhase);
+            }
+        }
+    }
+
+    private void innerRunFetch(ScoreDoc[] scoreDocs, int numShards, SearchPhaseController.ReducedQueryPhase reducedQueryPhase) {
+        ArraySearchPhaseResults<FetchSearchResult> fetchResults = new ArraySearchPhaseResults<>(numShards);
+        final List<Map<Integer, RankDoc>> rankDocsPerShard = false == shouldExplainRankScores(context.getRequest())
+            ? null
+            : splitRankDocsPerShard(scoreDocs, numShards);
+        final ScoreDoc[] lastEmittedDocPerShard = context.getRequest().scroll() != null
+            ? SearchPhaseController.getLastEmittedDocPerShard(reducedQueryPhase, numShards)
+            : null;
+        final List<Integer>[] docIdsToLoad = SearchPhaseController.fillDocIdsToLoad(numShards, scoreDocs);
+        final CountedCollector<FetchSearchResult> counter = new CountedCollector<>(
+            fetchResults,
+            docIdsToLoad.length, // we count down every shard in the result no matter if we got any results or not
+            () -> {
+                try (fetchResults) {
+                    moveToNextPhase(fetchResults.getAtomicArray(), reducedQueryPhase);
+                }
```

Comment:
This overall change makes sense to me. I dislike keeping an object around in the constructor and handling lifetimes like we did. Its always confusing.

Just calling out now that with this try (autoClose), we will now release the `fetchResults` AFTER `context.executeNextPhase` is called. I am not sure that is a big issue or not as generally the search phases do asynchronous actions.

Reply:
That's to some extend by design :D The aggressive closing of these things has been causing a leak (preexisting bug) to show at a higher rate. Let me deal with that bug in a separate PR today and then I'll reintroduce aggressive closing here again tomorrow or so.